x-pack/legacy/plugins/siem/public/components/paginated_table/index.tsx

```diff
@@ -0,0 +1,317 @@
+ */
+import {
+  EuiBasicTable,
+  EuiButtonEmpty,
+  EuiContextMenuItem,
+  EuiContextMenuPanel,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiPagination,
+  EuiPanel,
+  EuiPopover,
+} from '@elastic/eui';
+import { isEmpty, noop, getOr } from 'lodash/fp';
+import React, { memo, useState, useEffect } from 'react';
+import styled from 'styled-components';
+
+import { Direction } from '../../graphql/types';
+import { HeaderPanel } from '../header_panel';
+import { LoadingPanel } from '../loading';
+
+import * as i18n from './translations';
+
+export interface ItemsPerRow {
+  text: string;
+  numberOfRow: number;
+}
+
+export interface SortingBasicTable {
+  field: string;
+  direction: Direction;
+  allowNeutralSort?: boolean;
+}
+
+export interface Criteria {
+  page?: { index: number; size: number };
+  sort?: SortingBasicTable;
+}
+
+// Using telescoping templates to remove 'any' that was polluting downstream column type checks
+export interface BasicTableProps<T, U = T, V = T, W = T, X = T, Y = T, Z = T, AA = T, AB = T> {
+  columns:
+    | [Columns<T>]
+    | [Columns<T>, Columns<U>]
+    | [Columns<T>, Columns<U>, Columns<V>]
+    | [Columns<T>, Columns<U>, Columns<V>, Columns<W>]
+    | [Columns<T>, Columns<U>, Columns<V>, Columns<W>, Columns<X>]
+    | [Columns<T>, Columns<U>, Columns<V>, Columns<W>, Columns<X>, Columns<Y>]
+    | [Columns<T>, Columns<U>, Columns<V>, Columns<W>, Columns<X>, Columns<Y>, Columns<Z>]
+    | [
+        Columns<T>,
+        Columns<U>,
+        Columns<V>,
+        Columns<W>,
+        Columns<X>,
+        Columns<Y>,
+        Columns<Z>,
+        Columns<AA>
+      ]
+    | [
+        Columns<T>,
+        Columns<U>,
+        Columns<V>,
+        Columns<W>,
+        Columns<X>,
+        Columns<Y>,
+        Columns<Z>,
+        Columns<AA>,
+        Columns<AB>
+      ];
+  headerCount: number;
+  headerSupplement?: React.ReactElement;
+  headerTitle: string | React.ReactElement;
+  headerTooltip?: string;
+  headerUnit: string | React.ReactElement;
+  itemsPerRow?: ItemsPerRow[];
+  limit: number;
+  loading: boolean;
+  loadingTitle?: string;
+  loadPage: (activePage: number) => void;
+  onChange?: (criteria: Criteria) => void;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  pageOfItems: any[];
+  sorting?: SortingBasicTable;
+  totalCount: number;
+  updateActivePage: (activePage: number) => void;
+  updateLimitPagination: (limit: number) => void;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  updateProps?: { [key: string]: any };
+}
+
+export interface Columns<T> {
+  field?: string;
+  name: string | React.ReactNode;
+  isMobileHeader?: boolean;
+  sortable?: boolean;
+  truncateText?: boolean;
+  hideForMobile?: boolean;
+  render?: (item: T) => void;
+  width?: string;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const PaginatedTable = memo<BasicTableProps<any>>(
```

Comment:
@spong could you please help me figure out how to not do an `any` here? This was copied from `load_more_table`, but I'm not sure how to pass the arguments in that manner using `React.memo`

Reply:
Hey @stephmilovic , I find this question quite interesting, so had a try as well. 
Providing my implementation here (I did it for Hosts and Auths tables as example):
Adding these types:
```
export declare type AuthTableColumns = [
  Columns<AuthenticationsEdges>,
  Columns<AuthenticationsEdges>,
  Columns<AuthenticationsEdges>,
  Columns<AuthenticationsEdges>,
  Columns<AuthenticationsEdges>,
  Columns<AuthenticationsEdges>,
  Columns<AuthenticationsEdges>,
  Columns<AuthenticationsEdges>,
  Columns<AuthenticationsEdges>
];

export declare type HostsTableColumns = [
  Columns<string>,
  Columns<string>,
  Columns<string>,
  Columns<string>
];

declare type BasicTableColumns = AuthTableColumns | HostsTableColumns;

declare type SiemTables = BasicTableProps<BasicTableColumns>;
```

change the typeHolder for this interface:
```
export interface BasicTableProps<T> {
  columns: T;
  headerCount: number;
  headerSupplement?: React.ReactElement;
  headerTitle: string | React.ReactElement;
  headerTooltip?: string;
  headerUnit: string | React.ReactElement;
  itemsPerRow?: ItemsPerRow[];
  limit: number;
  loading: boolean;
  loadingTitle?: string;
  loadPage: (activePage: number) => void;
  onChange?: (criteria: Criteria) => void;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  pageOfItems: any[];
  sorting?: SortingBasicTable;
  totalCount: number;
  updateActivePage: (activePage: number) => void;
  updateLimitPagination: (limit: number) => void;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  updateProps?: { [key: string]: any };
}
```

and we can remove any here:
```
export const PaginatedTable = memo<SiemTables>(
```

Reply:
@angorayc you are the best <3